src/main/java/com/algorand/algosdk/transaction/MethodCallParams.java

```diff
@@ -40,6 +78,26 @@ public MethodCallParams(Long appID, Method method, List<Object> methodArgs, Stri
         this.flatFee = flatFee;
         this.rekeyTo = rekeyTo;
         this.signer = signer;
+        this.foreignAccounts = fAccounts;
+        this.foreignAssets = fAssets;
+        this.foreignApps = fApps;
+
+        Address senderAddr;
+        try {
+            senderAddr = new Address(this.sender);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("sender mnemonic cannot generate an address");
+        }
+        this.foreignAccounts.remove(senderAddr);
```

Comment:
This is a nice optimization, but I'm wondering if it might mess up the existing indices:

e.g. `foreignAccounts = [senderAddr, addr2, addr3]` -> `[addr2, addr3]` then if the smart contract tries to access `foreignAccounts[3]`, it could panic and the user might not know what might be wrong?

Reply:
This is an interesting example, ideally I think the foreign account array should not contain duplicated address, which means sender address should not be in there (even if it is not in the array but we still have an index for it).

My thoughts is that, the contract should not have access to index that lead to out of index error, since we populate the array index after we add the method call. I would like to leave this open for more discussion...

Reply:
My instinct is that this can be done at a lower level instead of at the SDK level, but I agree with your thoughts

Reply:
I think this optimization is ok to make, since foreign objects are not guaranteed to be in a specific order in the ABI. And the main point of specifying additional foreign objects, as far as I can see, is to make sure they are "available" to the AVM during execution (for inner txns and other opcodes which require foreign references). So removing duplicates and the sender if present does not prevent that usage.

Though I would strongly recommend documenting that the foreign object arrays in `MethodCallParams` represent _additional_ foreign objects that the app should be able to access during execution, and that there are no guarantees on the ordering of the additional objects, or even if every additional object will be explicitly included in the app call transaction.